crates/sui-core/src/authority.rs

```diff
@@ -455,14 +483,63 @@ impl AuthorityState {
     #[instrument(level = "debug", name = "process_cert_inner", skip_all)]
     async fn process_certificate(
         &self,
-        confirmation_transaction: ConfirmationTransaction,
-    ) -> Result<TransactionInfoResponse, SuiError> {
-        let certificate = confirmation_transaction.certificate;
+        tx_guard: CertTxGuard<'_>,
+        certificate: CertifiedTransaction,
+    ) -> SuiResult<TransactionInfoResponse> {
         let transaction_digest = *certificate.digest();
 
+        // The cert could have been processed by a concurrent attempt of the same cert, so check if
+        // the effects have already been written.
+        if self.database.effects_exists(&transaction_digest)? {
+            let info = self.make_transaction_info(&transaction_digest).await?;
+            debug!("Transaction {transaction_digest:?} already executed");
+            tx_guard.commit_tx();
+            return Ok(info);
+        }
+
+        // Errors originating from prepare_certificate may be transient (failure to read locks) or
+        // non-transient (transaction input is invalid, move vm errors). However, all errors from
+        // this function occur before we have written anything to the db, so we commit the tx
+        // guard and rely on the client to retry the tx (if it was transient).
+        let (temporary_store, signed_effects) = match self
+            .prepare_certificate(&certificate, transaction_digest)
+            .await
+        {
+            Err(e) => {
+                debug!("Error preparing transaction: {}", e);
+                tx_guard.commit_tx();
+                return Err(e);
+            }
+            Ok(res) => res,
+        };
+
+        // If commit_certificate returns an error, tx_guard will be dropped and the certificate
+        // will be persisted in the log for later recovery.
+        self.commit_certificate(temporary_store, &certificate, &signed_effects)
+            .await?;
+
+        // commit_certificate finished, the tx is fully committed to the store.
+        tx_guard.commit_tx();
+
+        Ok(TransactionInfoResponse {
+            signed_transaction: self.database.get_transaction(&transaction_digest)?,
+            certified_transaction: Some(certificate),
+            signed_effects: Some(signed_effects),
+        })
+    }
+
+    #[instrument(level = "debug", name = "prepare_certificate", skip_all)]
+    async fn prepare_certificate(
```

Comment:
Yes, it could be invoked earlier. This is an easy optimization that I think makes sense - it is very common (especially on full nodes, and probably on validators too once gossip is deployed) to receive many instances of the same tx concurrently. I know this happens a lot and is not just theoretical because it has exposed a few bugs. So, if we do this before the lock, we may be wasting CPU recomputing the same effects many times.